plugins/meta/vrf/main.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+// VRFNetConf represents the vrf configuration.
+type VRFNetConf struct {
+	types.NetConf
+
+	// VRFName is the name of the vrf to add the interface to.
+	VRFName string `json:"vrfname"`
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.PluginSupports("0.4.0"), bv.BuildString("vrf"))
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, result, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("missing prevResult from earlier plugin")
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+
+		if _, ok := err.(netlink.LinkNotFoundError); ok {
+			vrf, err = createVRF(conf.VRFName)
+		}
+
+		if err != nil {
+			return err
+		}
```

Comment:
should we check first for the erro and then if it is a LinkNotFound error?
```
		if err != nil {
		     if _, ok := err.(netlink.LinkNotFoundError); ok {
			vrf, err = createVRF(conf.VRFName)
                      }	 else {
                         return err
                     }
             }

Reply:
I don't like extra nesting, and the above version is behaviourally identical, as the type check will fail if err is nil.

But if you think this other version is more clear I can change.

Reply:
I was more curious about what happens if err=nil with this?
```err.(netlink.LinkNotFoundError);```
from your comments I assume is not a problem

Reply:
you are right, there is lot of nesting and is very ugly to read :)